vignettes/glue.Rmd

```diff
@@ -1,140 +1,206 @@
+x <- glue("
+  blah
+  ")
+unclass(x)
+
+# both a leading and trailing newline
+y <- glue("
+
+  blah
 
-  leading or trailing newlines can be added explicitly
-  as long as the trailing line contains at least one whitespace character
-   
   ")
+unclass(y)
 ```
 
-You can use `\\` at the end of a line continues to continue a single
-line:
+## Delimiters
+
+By default, code to be evaluated goes inside `{}` in a glue string.
+If want a literal curly brace in your string, double it:
 
 ```{r}
-glue("
-    A formatted string \\
-    can also be on a \\
-    single line
-    ")
+glue("The name of the package is {name}, not {{name}}.")
 ```
 
+Sometimes it's just more convenient to use different delimiters altogether, especially if the template text comes from elsewhere or is subject to external requirements.
+Consider this example where we want to interpolate the function name into a code snippet that defines a function:
 
-## Glue in pipes and pipelines
+```{r}
+fn_def <- "
+  <<NAME>> <- function(x) {
+    # imagine a function body here
+  }"
+glue(fn_def, NAME = "my_function", .open = "<<", .close = ">>")
+```
+
+In this glue string, `{` and `}` have very special meaning.
+If we forced ourselves to double them, suddenly it doesn't look like normal R code anymore.
+Using alternative delimiters is a nice option in cases like this.
+
+## Where glue looks for values
+
+By default, `glue()` evaluates the code inside `{}` in the caller environment:
+
+```{r, eval = FALSE}
+glue(..., .envir = parent.frame())
+```
 
-`glue_data()` is useful with magrittr pipes.
+So, for a top-level `glue()` call, that means the global environment.
 
 ```{r}
-`%>%` <- magrittr::`%>%`
-head(mtcars) %>% glue_data("{rownames(.)} has {hp} hp")
+x <- "the caller environment"
+glue("By default, `glue()` evaluates code in {x}.")
 ```
-`glue()` is useful within dplyr pipelines.
 
-```{r,message=FALSE}
-library(dplyr)
-head(iris) %>%
-  mutate(description = glue("This {Species} has a petal length of {Petal.Length}"))
+But you can provide more narrowly scoped values by passing them to `glue()` in `name = value` form:
+
+```{r}
+x <- "the local environment"
+glue(
+  "`glue()` can access values from {x} or from {y}. {z}",
+  y = "named arguments",
+  z = "Woo!"
+)
+```
+
+If the relevant data lives in a data frame (or list or environment), use `glue_data()` instead:
+
+```{r}
+mini_mtcars <- head(cbind(model = rownames(mtcars), mtcars))
+rownames(mini_mtcars) <- NULL
+glue_data(mini_mtcars, "{model} has {hp} hp.")
+```
+
+`glue_data()` is very natural to use with the pipe:
+
+```{r, eval = getRversion() >= "4.1.0"}
+mini_mtcars |>
+  glue_data("{model} gets {mpg} miles per gallon.")
 ```
 
+Returning to `glue()`, recall that it defaults to evaluation in the caller environment.
+This has happy implications inside a `dplyr::mutate()` pipeline.
+The data-masking feature of `mutate()` means the columns of the target data frame are "in scope" for a `glue()` call:
+
+```r
+library(dplyr)
```

Comment:
This has to be a fixed chunk, given the use of dplyr, which is a drag. But I want to show this.